### ISSUE TITLE
How to include the installation parameters and make BLAS arguments in the requirements.txt?

### ISSUE THREAD

**Issue by AayushSameerShah** [abetlen/llama-cpp-python#989]
### Discussed in https://github.com/abetlen/llama-cpp-python/discussions/976

<div type='discussions-op-text'>

<sup>Originally posted by **AayushSameerShah** December  6, 2023</sup>
# 🙋🏻‍♂️ 
I am building a simple project in my local machine **and then will deploy on AWS ubuntu server as an API**.
👉🏻 In the development environment I have installed this application using:
```sh
$env:CMAKE_ARGS = "-DLLAMA_BLAS=ON -DLLAMA_BLAS_VENDOR=OpenBLAS"
pip install llama-cpp-python
```
👉🏻 When I do the `pip freeze > requirements.txt` *(obviously)* it is not going to include these env arguments.

### ⁉ Question
*How can In include these arguments in the requirements so that in the deployment environment I don't have to configure things manually?* - **And is that even possible?**</div>

**Comment by abetlen** [comment]
@AayushSameerShah I don't believe that's possible and yes I agree it should be, the only option is to maybe publish multiple packages for each configuration or guess at the best config per platform.

**Comment by ceruleandeep** [comment]
Please try and figure something out here :)

Edit: I wrote the comments below before realising that llama-cpp-python does provide prebuilt wheels, lots of them, and it might be possible for my installer to figure out the right one and install it. I will try it.

--

I have a [tiny project](https://github.com/ceruleandeep/ComfyUI-LLaVA-Captioner) that uses llama-cpp-python, and it will stay tiny because no one can get it to install on Windows. Meanwhile [llamafile](https://github.com/Mozilla-Ocho/llamafile) has somehow produced a llama.cpp binary that works as-is on 6 platforms including MPS and CUDA. And for reasons I can't work out yet, inference with llamafile is 2x faster than with llama-cpp-python on my 4090 on Linux.  

It's probably far-fetched, but is there some way to use llamafile as a backend for llama-cpp-python? Most of the potential users of my tiny project are on Windows, where llamafile is at its best and llama.cpp is at its worst in terms of installation. 

Or, I just noticed that llama.cpp provides prebuilt binaries in their releases. Can these be dropped in?

**Comment by abetlen** [comment]
@AayushSameerShah late reply but I have a bettter answer for you! In newer versions of `pip` it's possible to specify `-C/--config-settings` flag which passes args to the build tool (scikit-build-core in our case) so you can do

```bash
# requirements.txt
llama-cpp-python -C cmake.args="-DLLAMA_BLAS=ON;-DLLAMA_BLAS_VENDOR=OpenBLAS"
```

This is also documented now in the README.md now under `CLI / requirements.txt`. You may need to run `pip install --upgrade pip` if you get an error about the `-C` option to upgrade to a newer version of pip.

Hope this helps.
